### DISCUSSION TITLE
What does the plus symbol do here?

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#8927] [General]
I found a query online and it does a `edges+` query:

```
from TargetMethod end, MyMainMethod entryPoint
where edges+(entryPoint, end)
select end, entryPoint, end, "Found a path from start to target."
```

What does the `+` symbol do here?

**Comment by MathiasVP** [comment]
Hi @CaledoniaProject,

The `+` symbol means "transitive closure". That is, the query you've posted finds all `end, entryPoint` pairs where there exists a sequence of nodes `n1, ..., nk` such that `edges(entryPoint, n1), edges(n1, n2), ..., edges(nk, end)`.

You can read about transitive closures here: https://codeql.github.com/docs/ql-language-reference/recursion/#transitive-closures

**Comment by CaledoniaProject** [comment]
How can I do this in C++? I'm just trying to find calls from main() to system():

```
/**
 * @name test
 * @kind path-problem
 */

import cpp
import DataFlow::PathGraph
import semmle.code.cpp.dataflow.DataFlow
import semmle.code.cpp.dataflow.TaintTracking
import semmle.code.cpp.valuenumbering.GlobalValueNumbering
import semmle.code.cpp.controlflow.Guards

class MainFunction extends Function {
  MainFunction() {
      getName() = "main"
  }
}

class SystemFunction extends Function {
  SystemFunction() { 
      getName() = "system"
  }
}

from MainFunction start, SystemFunction end
where edges+(start, end)
select start, end, "Found a path from start to target."
```

**Reply by MathiasVP** [reply]
The following query should do that:
```ql
/**
 * @name test
 * @kind path-problem
 */

import cpp

class MainFunction extends Function {
  MainFunction() { getName() = "main" }
}

class SystemFunction extends Function {
  SystemFunction() { getName() = "system" }
}

query predicate edges(Function caller, Function callee) { caller.calls(callee) }

from MainFunction start, SystemFunction end
where edges+(start, end)
select end, start, end, "Found a path from start to target."
```
Note that this will ignore calls through function pointers and inheritance. If you want to include some of those cases you could make the `edges` predicate a bit smarter and use our `ResolveCall` library:
```ql
/**
 * @name test
 * @kind path-problem
 */

import cpp
import semmle.code.cpp.ir.dataflow.ResolveCall

class MainFunction extends Function {
  MainFunction() { getName() = "main" }
}

class SystemFunction extends Function {
  SystemFunction() { getName() = "system" }
}

query predicate edges(Function caller, Function callee) {
  callee = resolveCall(any(Call call | call.getEnclosingFunction() = caller))
}

from MainFunction start, SystemFunction end
where edges+(start, end)
select end, start, end, "Found a path from start to target."
```
However, if you're going down that route, you should take a look at our [dataflow library](https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-cpp/) which is well suited for these kinds of things.

**Comment by CaledoniaProject** [comment]
It does not work with this cpp file: [test.txt](https://github.com/github/codeql/files/8595432/test.txt)

I'm just trying to find all calls to system() initiated from main(), the query should find 3 call paths, but it didn't find any.

**Reply by CaledoniaProject** [reply]
@MathiasVP

**Reply by intrigus-lgtm** [reply]
What did you try, when debugging the problem?

**Reply by CaledoniaProject** [reply]
I combined both solutions to test it, quick evaluation works fine, but the main query finds nothing.

```
/**
 * @name test
 * @kind path-problem
 */

import cpp
import semmle.code.cpp.ir.dataflow.ResolveCall

class MainFunction extends Function {
  MainFunction() { getName() = "main" }
}

class SystemFunction extends Function {
  SystemFunction() { getName() = "system" }
}

query predicate edges1(Function caller, Function callee) {
  callee = resolveCall(any(Call call | call.getEnclosingFunction() = caller))
}
query predicate edges2(Function caller, Function callee) { caller.calls(callee) }


from MainFunction start, SystemFunction end
where edges1+(start, end) or edges2+(start, end)
select end, start, end, "Found a path from start to target."
```

**Reply by MathiasVP** [reply]
Note that the query predicate _must_ be named `edges` (i.e., not `edges1` or `edges2`). Does that help?

**Reply by CaledoniaProject** [reply]
I've created a wrapper predicate and it's still not working:

```
/**
 * @name test
 * @kind path-problem
 */

// https://github.com/github/codeql/discussions/8927#discussioncomment-2661524

import cpp
import semmle.code.cpp.ir.dataflow.ResolveCall

class MainFunction extends Function {
  MainFunction() { getName() = "main" }
}

class SystemFunction extends Function {
  SystemFunction() { getName() = "system" }
}

predicate edges1(Function caller, Function callee) {
  callee = resolveCall(any(Call call | call.getEnclosingFunction() = caller))
}
predicate edges2(Function caller, Function callee) { caller.calls(callee) }

query predicate edges(Function caller, Function callee) {
  edges1(caller, callee) or edges2(caller, callee)
}

from MainFunction start, SystemFunction end
where edges+(start, end)
select end, start, end, "Found a path from start to target."
```

**Reply by CaledoniaProject** [reply]
Hi @MathiasVP, can you help?

**Reply by MathiasVP** [reply]
Ah, yes. I think see what's going on now. Most likely, the issue isn't with your QL code, but instead how the path is rendered.

If you look in the database folder you'll find a `codeql-database.yml` file containing a `sourceLocationPrefix` field. CodeQL will only show the path when the endpoints are located in a file that's nested within that directory. In this case, the `system` function is most likely located in your standard library's directory.

There are a couple of ways you can go about fixing this:
- As a quick fix, you can modify the `sourceLocationPrefix` value to something like "/" to not exclude any paths based on location. That should cause your paths to show up, but it might mean that the elements interact weirdly with the CLI or with VSCode (since you're editing a field that we don't expect you to edit).
- A more proper fix is to specify another working directory using `codeql database create ... --working-dir=my-directory` where `my-directory` is a directory that's both a root of your standard library's included files and your own project files.

**Reply by CaledoniaProject** [reply]
system() is defined in libc and I don't have its source code ...

**Reply by MathiasVP** [reply]
You don't need the source code for the `system` function, but you need to specify a `--working-dir` that is a prefix of the directory where the `system` function is _declared_`. For instance, my `cstdlib` header is located at `/Applications/Xcode-12.5.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstdlib`, and if I specify `codeql database create ... --working-dir=/Applications` any path that ends with an element in a file that is within the `/Applications` folder will be shown.

Alternatively, if your program is sufficiently simple to compile (which the example program you linked is) you can `cd` into `/Applications` and run `codeql database create my-test-db --language=cpp --command="clang++ absolute-path-to-test-file.cpp"`. That should give you a database with a `sourceLocationPrefix` that will show paths even though they end inside a of libc folder.
